crates/telio-firewall/src/firewall.rs

```diff
@@ -704,29 +743,113 @@ impl StatefullFirewall {
         };
 
         let identifier_sequence = {
-            let payload = icmp_packet.payload();
-            if payload.len() < 4 {
-                return None;
-            }
-            #[allow(index_access_check)]
-            let bytes = payload[0..4].try_into();
-            if let Ok(bytes) = bytes {
-                u32::from_ne_bytes(bytes)
-            } else {
-                return None;
-            }
+            let bytes = icmp_packet
+                .payload()
+                .get(0..4)
+                .and_then(|b| b.try_into().ok());
+            let bytes = unwrap_option_or_return!(bytes, None);
+            u32::from_ne_bytes(bytes)
         };
         let (remote_addr, local_addr) = if inbound {
             (ip.get_source().into(), ip.get_destination().into())
         } else {
             (ip.get_destination().into(), ip.get_source().into())
         };
-        Some(IcmpConn {
+        let key = IcmpConn {
             remote_addr,
             local_addr,
             response_type,
             identifier_sequence,
-        })
+        };
+        Some(IcmpKey::IcmpConn(key))
+    }
+
+    fn build_icmp_error_key(icmp_packet: IcmpPacket, is_v4: bool) -> Option<IcmpKey> {
+        let inner_packet = unwrap_option_or_return!(icmp_packet.payload().get(4..), None);
+        let icmp_error_key = if is_v4 {
+            let packet = unwrap_option_or_return!(Ipv4Packet::try_from(inner_packet), None);
+            match packet.get_next_level_protocol() {
+                IpNextHeaderProtocols::Udp => {
+                    IcmpErrorKey::Udp(Self::build_udp_key(&packet, false))
+                }
+                IpNextHeaderProtocols::Tcp => {
+                    IcmpErrorKey::Tcp(Self::build_tcp_key(&packet, false).map(|(key, _)| key))
+                }
+                IpNextHeaderProtocols::Icmp => IcmpErrorKey::Icmp(
+                    Self::build_icmp_key(&packet, false).and_then(|key| match key {
+                        IcmpKey::IcmpConn(key) => Some(key),
+                        IcmpKey::Error { .. } => None,
+                    }),
+                ),
```

Comment:
Seems to me that all udp packets that fail to parse will have the same key. Similarly for the tcp packets (and possibly icmp). I would expect that it would be an error of some kind, not just an ordinary key value.

That is why I was expecting the `IcmpErrorKey` variants not to contain options.

Reply:
Technically yes but practically no. Failing to parse a packet that is nested in an ICMP error shouldn't really produce an error, it just signals that we should reject the packet. If we can't parse a packet the key becomes `None`, which is automatically rejected.

Reply:
You are right but that made me realize that the key is not the key and I was thinking all this time tha it is and it ends up the the hashmap. As far as I can see `IcmpKey` should be an alias to `Result<IcmpConn, IcmpError>` where `IcmpError` is renamed `IcmpErrorKey`.

Reply:
Good idea. I simplified some other stuff as well to rely on `IcmpErrorKey::None` to simplify `Option<IcmpKey>` to just `IcmpKey` as well.